File: src/lib/mockSuppliers.ts
// Stand-ins for the suppliers' own ordering APIs (the whiteboard's "Home Depot API").
// Each supplier answers a quote with price, stock and lead time, and accepts an order.
// Answers are deterministic per supplier, part and day, and about 1 call in 12 fails with a
// 503 so the worker's retry path is exercised.

import { q, one } from "./db";
import { AS_OF } from "./config";
import { addDays } from "./dates";
import { SUPPLIERS } from "./catalog";

function hash(s: string) {
  let h = 2166136261;
  for (let i = 0; i < s.length; i++) h = Math.imul(h ^ s.charCodeAt(i), 16777619);
  return (h >>> 0) / 4294967296;
}

export function supplierBySlug(slug: string) {
  return SUPPLIERS.find((s) => s.slug === slug.toLowerCase()) ?? null;
}

export type Quote = {
  supplier: string;
  sku: string;
  quantity: number;
  available: number;
  unitPrice: number;
  leadDays: number;
  promisedDate: string;
  canFill: boolean;
};

export class SupplierUnavailable extends Error {}

export async function quote(slug: string, sku: string, quantity: number, attemptSalt = ""): Promise<Quote | null> {
  const s = supplierBySlug(slug);
  if (!s) return null;
  if (hash(`${slug}|${sku}|${attemptSalt}|${Date.now() >> 12}`) < 0.08) throw new SupplierUnavailable(`${s.code} API returned 503`);
  const row = await one<{ unit_price: number; nominal_lead_days: number }>(
    `SELECT unit_price, nominal_lead_days FROM part_suppliers WHERE sku=$1 AND supplier=$2`,
    [sku, s.code],
  );
  if (!row) return null;
  const r = hash(`${slug}|${sku}|stock`);
  const available = Math.round(80 + r * 900);
  const leadDays = row.nominal_lead_days + (available >= quantity ? 0 : 14);
  return {
    supplier: s.code,
    sku,
    quantity,
    available,
    unitPrice: Math.round(row.unit_price * (0.97 + hash(`${slug}|${sku}|px`) * 0.06) * 100) / 100,
    leadDays,
    promisedDate: addDays(AS_OF, leadDays),
    canFill: available >= quantity,
  };
}

export async function placeOrder(slug: string, sku: string, quantity: number) {
  const qt = await quote(slug, sku, quantity, "order");
  if (!qt) return null;
  const ref = `${slug.toUpperCase()}-${Math.floor(hash(`${slug}|${sku}|${quantity}|${Date.now()}`) * 1e8)
    .toString()
    .padStart(8, "0")}`;
  return { ...qt, externalRef: ref, accepted: true };
}

export async function supplierSlugsForSku(sku: string) {
  const rows = await q<{ supplier: string }>(`SELECT supplier FROM part_suppliers WHERE sku=$1`, [sku]);
  return rows.map((r) => SUPPLIERS.find((s) => s.code === r.supplier)!.slug);
}

File: backend/supply/mock_suppliers.py
"""Stand-ins for the suppliers' own ordering APIs (the whiteboard's "Home Depot API").

Each supplier answers a quote with price, stock and lead time, and accepts an order. Answers
are deterministic per supplier and part, and about 1 call in 12 fails with a 503 so the
worker's retry path is exercised.

Placing an order honours an Idempotency-Key header the way a real supplier API does: the
supplier stores the key with the result, and a repeat with the same key gets the stored result
back instead of a second order. mock_supplier_orders is that store.

MOCK_SUPPLIER_LATENCY_MS makes the order call take that long before the supplier records the
order, like a slow supplier. It is 0 unless set; it gives the kill test a window to stop the
worker while a supplier call is in flight.
"""

import json
import os
import time
import uuid

from sqlalchemy import text

from .catalog import SUPPLIERS
from .config import as_of
from .dates import add_days
from .db import engine, one, rows
from .prng import js_round


class SupplierUnavailable(Exception):
    pass


class KeyReused(Exception):
    pass


def _hash(s: str) -> float:
    """FNV-1a, 32-bit, scaled to [0, 1)."""
    h = 2166136261
    for ch in s:
        h = ((h ^ ord(ch)) * 16777619) & 0xFFFFFFFF
    return h / 4294967296


def supplier_by_slug(slug: str) -> dict | None:
    return next((s for s in SUPPLIERS if s["slug"] == slug.lower()), None)


def quote(slug: str, sku: str, quantity: int, attempt_salt: str = "") -> dict | None:
    s = supplier_by_slug(slug)
    if not s:
        return None
    if _hash(f"{slug}|{sku}|{attempt_salt}|{int(time.time() * 1000) >> 12}") < 0.08:
        raise SupplierUnavailable(f"{s['code']} API returned 503")
    row = one("SELECT unit_price, nominal_lead_days FROM part_suppliers WHERE sku=:sku AND supplier=:s", {"sku": sku, "s": s["code"]})
    if not row:
        return None
    available = js_round(80 + _hash(f"{slug}|{sku}|stock") * 900)
    lead_days = row["nominal_lead_days"] + (0 if available >= quantity else 14)
    return {
        "supplier": s["code"],
        "sku": sku,
        "quantity": quantity,
        "available": available,
        "unitPrice": js_round(row["unit_price"] * (0.97 + _hash(f"{slug}|{sku}|px") * 0.06) * 100) / 100,
        "leadDays": lead_days,
        "promisedDate": add_days(as_of(), lead_days),
        "canFill": available >= quantity,
    }


def place_order(slug: str, sku: str, quantity: int, idempotency_key: str | None) -> tuple[str | None, bool]:
    """Place an order. Returns (response JSON text, replayed). A known key replays the stored text."""
    s = supplier_by_slug(slug)
    if not s:
        return None, False
    key = idempotency_key or f"unkeyed-{uuid.uuid4()}"
    stored = one("SELECT supplier, sku, quantity, response FROM mock_supplier_orders WHERE idempotency_key = :k", {"k": key})
    if stored:
        if (stored["supplier"], stored["sku"], stored["quantity"]) != (s["code"], sku, quantity):
            raise KeyReused(f"Idempotency-Key {key} was already used for a different order")
        return stored["response"], True
    qt = quote(slug, sku, quantity, "order")
    if not qt:
        return None, False
    time.sleep(int(os.environ.get("MOCK_SUPPLIER_LATENCY_MS") or 0) / 1000)
    ref = f"{slug.upper()}-{int(_hash(f'{slug}|{sku}|{quantity}|{time.time_ns()}') * 1e8):08d}"
    result = json.dumps({**qt, "externalRef": ref, "accepted": True})
    with engine().begin() as c:
        inserted = c.execute(
            text(
                """INSERT INTO mock_supplier_orders (idempotency_key, supplier, sku, quantity, response)
                   VALUES (:k, :s, :sku, :q, :r) ON CONFLICT (idempotency_key) DO NOTHING RETURNING 1"""
            ),
            {"k": key, "s": s["code"], "sku": sku, "q": quantity, "r": result},
        ).first()
    if not inserted:
        # Another request with this key won the race: return what it stored.
        return place_order(slug, sku, quantity, key)
    return result, False


def supplier_slugs_for_sku(sku: str) -> list[str]:
    codes = [r["supplier"] for r in rows("SELECT supplier FROM part_suppliers WHERE sku=:sku ORDER BY supplier", {"sku": sku})]
    return [s["slug"] for s in SUPPLIERS if s["code"] in codes]
